modules/schema_validation.lua

```diff
@@ -35,315 +26,264 @@ local function dump(o)
+    end
+  end
+  return true
+end
 
-
-  local function nodeVerify(xmlNode, dataNode, key)
-    if (xmlNode.class == 'enum') then
-      if type(dataNode) == "table" then
-        local result
-        result,errorMessage[ key ] = compare_table_key(types.enum[xmlNode['type'] ], dataNode, true)
-        bool_result = result and bool_result
-      elseif (types.enum[xmlNode['type'] ][dataNode] and xmlNode.array == 'true') then
-        bool_result = false
-        errorMessage[ key ] = "expected ENum : '" .. key .."' with type [ array ]"
-      elseif (not types.enum[xmlNode['type'] ][dataNode]) then
-        bool_result = false
-        errorMessage[ key ] = "expected: '" .. key .."' with type [ Enum ]"
+local function CheckValue( data_elem, schemaElem, data_name )
+  if schemaElem["minlength"] ~= nil then
+    if(data_elem < schemaElem["minlength"]) then
+        return false, "For parameter "..data_name.." get size: "..data_elem..", expected minvalue:"..schemaElem["minvalue"]
+    end
+  end
+  if schemaElem["maxlength"] ~= nil then
+      if(data_elem > schemaElem["maxlength"]) then
+        return false, "For parameter "..data_name.." get size: "..data_elem..", expected maxvalue:"..schemaElem["maxvalue"]
       end
-    elseif (xmlNode.class == 'struct') then
-      if (type(dataNode) == 'table') then
-        local result
-        -- print("Parse : " .. dump(dataNode))
-        -- print(dump(types.struct[xmlNode['type'] ]))
-        result, errorMessage[ key ] = compare_table_key(types.struct[xmlNode['type'] ], dataNode)
-        bool_result = bool_result and result
-      elseif(types.struct[xmlNode['type'] ][dataNode] and xmlNode.array == 'true') then
-        bool_result = false
-        errorMessage[ key ] = "expected: '" .. key .."' with type [ Struct ]"
-      elseif(not types.struct[xmlNode['type'] ][dataNode]) then
-         bool_result = false
-         errorMessage[ key ] = "expected struct: '" .. key .."' with type [ array ]"
+  end
+  return true
+end
+
+function module.mt.__index:CompareType(data_elem, schemaElem, isArray, nameofParameter)
+  local elem1 = type(data_elem)
+  if (isArray=='true') then
+    return self:CheckTypesInArray(data_elem, schemaElem, nameofParameter)
+  else -- if element is not array
+    if (schemaElem == 'Integer' or schemaElem == 'Float') then 
+      if elem1 == 'number'  then
+        return CheckValue( data_elem, schemaElem, nameofParameter )
+      else 
+        return false, "Parameter ".. nameofParameter..": got "..elem1..", expected "..schemaElem
+      end    
+    elseif schemaElem == 'String' then 
+      if elem1 == 'string' then
+        return CheckLength( data_elem, schemaElem, nameofParameter )
+      else 
+        return false, "Parameter ".. nameofParameter..": got "..elem1..", expected "..schemaElem
+      end 
+    elseif schemaElem == 'Boolean' then
+      if elem1 == 'boolean' then
```

Comment:
@VVeremjova  Please try to avoid if/else if you use return statements. I think it would be euisier to read